src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,190 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    private static class Node<T> {
+        private T item;
+        private Node<T> next;
+        private Node<T> prev;
+
+        public Node(Node<T> prev, T item, Node<T> next) {
+            this.prev = prev;
+            this.item = item;
+            this.next = next;
+        }
+    }
+
     @Override
     public boolean add(T value) {
-        return false;
+        if (head == null) {
+            head = new Node<>(null, value, null);
+            tail = new Node<>(null, value, null);
+        } else {
+            if (head.next == null) {
+                tail = new Node<>(head, value, null);
+                head.next = tail;
+            } else {
+                Node<T> newNode = new Node<>(tail, value, null);
+                tail.next = newNode;
+                tail = newNode;
+            }
+        }
+        size++;
+
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIsIndexExist(index, size, "Invalid index");
+
+        if (index == size) {
+            add(value);
+            return;
+        }
 
+        if (index == 0) {
+            addAtBeginning(value);
+        } else {
+            addByIndex(value, index);
+        }
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        int count = 0;
+        Node<T> searchNode = head;
+        while (count < index) {
+            searchNode = searchNode.next;
+            count++;
+        }
+
+        return searchNode;
+    }
+
+    private void addByIndex(T value, int index) {
+        Node<T> nodeByIndex = findNodeByIndex(index);
+        Node<T> newNode = new Node<>(nodeByIndex.prev, value, nodeByIndex);
+        nodeByIndex.prev.next = newNode;
+        nodeByIndex.prev = newNode;
+        size++;
+    }
+
+    private void addAtBeginning(T value) {
+        Node<T> tempNode = new Node<>(null, value, head);
+        head.prev = tempNode;
+        head = tempNode;
+        size++;
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (T listItem: list) {
+            add(listItem);
+        }
+
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIsIndexExist(index, size - 1, "Invalid index");
+
+        return findNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIsIndexExist(index, size - 1, "Invalid index");
+
+        Node<T> nodeByIndex = findNodeByIndex(index);
+        T valueBeforeSet = nodeByIndex.item;
+        nodeByIndex.item = value;
+
+        return valueBeforeSet;
     }
 
     @Override
     public T remove(int index) {
```

Comment:
create method unlink that will look like this and code will be simplified 
  private T unlink(Node<T> targetNode) {
        final T element = targetNode.item;
        Node<T> prev = targetNode.prev;
        Node<T> next = targetNode.next;

        if (prev == null) {
            first = next;
        } else {
            prev.next = next;
            targetNode.prev = null;
        }

        if (next == null) {
            last = prev;
        } else {
            next.prev = prev;
            targetNode.next = null;
        }

        size--;
        return element;
    }